flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/HistogramBasedWatermarkEmitter.java

```diff
@@ -0,0 +1,263 @@
+		super.open(parameters);
+		inSampling = true;
+		setNextSamplingTimer(true);
+	}
+
+	@Override
+	public Watermark getCurrentWatermark() {
+
+		// we do not risk overflow because we always subtract from the currentMaxSeenTimestamp
+		// we do not risk underflow because the allowed lateness will always be smaller than the
+		// currentMaxSeenTimestamp and bigger than Long.MIN_VALUE
+		long nextWmTimestamp = currentMaxSeenTimestamp - allowedLateness;
+		return new Watermark(nextWmTimestamp);
+	}
+
+	@Override
+	public long extractTimestamp(T element, long previousElementTimestamp) {
+		long newTimestamp = extractTimestamp(element);
+		long lateness = currentMaxSeenTimestamp - newTimestamp;
+
+		if (currentMaxSeenTimestamp == Long.MIN_VALUE || lateness <= 0) {
+			lateness = 0;
+			currentMaxSeenTimestamp = newTimestamp;
+		}
+
+		if (inSampling) {
+			int bucket = computeBucket(lateness);
+			Integer counter = histBuckets.get(bucket);
+			if (counter == null) {
+				counter = 0;
+			}
+			histBuckets.put(bucket, counter + 1);
+			noOfEvents++;
+		}
+		return newTimestamp;
+	}
+
+	@Override
+	public void trigger(long timestamp) throws Exception {
+		processSamplingTimer();
+	}
+
+	/**
+	 * Adds the next sampling timer and does the necessary processing, depending on if we
+	 * transition from a sampling to a non-sampling period, or the other way round.
+	 * More specifically, if we transition from a sampling period to a non-sampling one, then
+	 * it updates the {@link #allowedLateness} and resets all the remaining book-keeping
+	 * data structures to their initial values. The interval till the next timer depends
+	 * on if we enter, or we exit from a sampling period.
+	 */
+	private void processSamplingTimer() {
+		if (inSampling && !histBuckets.isEmpty()) {
+			// we transitioned from sampling to a non-sampling period
+			// so update the allowed lateness to reflect the new data,
+			// and set everything else its initial state for the next
+			// sampling period.
+			updateAllowedLateness();
+			restoreToInit();
+		}
+		inSampling = !inSampling;
+		setNextSamplingTimer(inSampling);
+	}
+
+	/**
+	 * Registers a timer related to the start or the end of a sampling period.
+	 * The timer can be either to start or to stop a sampling period. This is specified
+	 * with the {@code timerToStartSampling} flag. Based on this flag, this {@code inSampling}
+	 * flag is set to {@code true} or {@code false}, and a timer for the beginning or end of a
+	 * sampling period is added. The time for the next trigger to fire adjusts accordingly.
+	 *
+	 * @param enterSampling A flag indicating if the timer to register is to signal the
+	 *                      start of the next sampling period, or the end of the current one.
+	 */
+	private void setNextSamplingTimer(boolean enterSampling) {
+		long timeToNextSamplingTimer = System.currentTimeMillis() +
+			(enterSampling ? samplingPeriod : nonSamplingInterval);
+		if(!testing) {
```

Comment:
I think you can avoid this hack by passing a testing StreamingRuntimeContext which ignores the registerTimer() call.